### PR TITLE
[SPARK-3847] Prohibit Java enums from being used as partitioning keys; strengthen prohibitions on array keys in PairRDDFuncions

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch prevents Java enums from being used as keys for PairRDD partitioning operations.  Java enums' hashCodes are based on the enumerations' ordinal() values rather than the default Enum.hashCode, which produces machine-dependent values that cannot safely be compared across machines.  We could add some special-case logic to HashPartitioner for handling this case, but I've chosen not to do that because it may give users the false impression that it's safe to use Java enums as components of a partitioning key, e.g. `PairRDD[(Enum, Enum), T]`.

In addition, this patch strengthens some of the existing error-checking which prohibits arrays from being used as keys for the PairRDDFunctions transformations.  [Previously](https://github.com/mesos/spark/pull/348), I added logic to throw exceptions if the default HashPartitioner was being used with array keys, since Java arrays' hashCodes are based on the arrays' object identities and not their contents.  However, this still permitted arrays to be used as keys if a custom partitioner was supplied.  This could still lead to errors, though, since we might call the default array hashCode when using arrays as keys in aggregation maps.  Therefore, I've strengthened the error-checking to always throw exceptions if arrays are being used in most PairRDDFunctions methods.

In case we have to add similar error checks in the future, I've factored the logic into a helper function in the `Partitioner` object.  I also improved the warning messages to reference the relevant JIRA issues.

### CODE REVIEW

**Comment by JoshRosen** [comment]
@aarondav @pwendell Does one of you want to do the final sign-off on this?

File: core/src/main/scala/org/apache/spark/rdd/OrderedRDDFunctions.scala [hunk 1]
```diff
@@ -72,6 +72,8 @@ class OrderedRDDFunctions[K : Ordering : ClassTag,
    * because it can push the sorting down into the shuffle machinery.
    */
   def repartitionAndSortWithinPartitions(partitioner: Partitioner): RDD[(K, V)] = {
+    val keyClass = implicitly[ClassTag[K]].runtimeClass
+    Partitioner.assertHashCodeIsWellBehaved(keyClass)
```
Comment (aarondav) [inline core/src/main/scala/org/apache/spark/rdd/OrderedRDDFunctions.scala]: Technically, this method may work even though the hash code is not well-behaved, if the partitioner is not a HashPartitioner, right?

Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/rdd/OrderedRDDFunctions.scala]: I think the concern here is that there's other code which might call `hashCode` on the items, so even if we use a custom partitioner we will still get incorrect results for Java arrays.  I think that this is safe as long as we can guarantee that the sorting phase doesn't call `hashCode`.

**Comment by JoshRosen** [comment]
Based on some offline with @aarondav, I'm going to re-work this to not allow Enums to be used, since the current behavior in this PR is misleading: we can automatically do the right thing for Enums used directly as keys, but cases like pairs of enums will still give incorrect output.  Throwing a warning for the cases that we can safely detect is a good opportunity to educate users.  I'll also add some loud warnings to the documentation if I can find a good place.

**Comment by andrewor14** [comment]
As discussed offline, we will extend the existing checks against using Arrays as keys to using Enums as well, instead of trying to manually fix this specific case. Otherwise we may give the user a false sense of hope if they later try to use, say, tuples of Enums as keys, which won't work for the same reason.

**Comment by JoshRosen** [comment]
I don't exactly have time to work on this right now, since it's going to be a really time consuming process to figure out all of the corner-cases and make sure that we're not breaking things.  I'm going to close this PR for now and will unassign the JIRA in case anyone else wants to take over.

**Pull request closed without merging**